### PR TITLE
FIX: don't allow category and tag tracking settings on staged users

### PR DESCRIPTION
**Pull request by nlalonde**
Configuring staged users to watch categories and tags is a way to sign
them up to get many emails. These emails may be unwanted and get marked
as spam, hurting the site's email deliverability.
Users can opt-in to email notifications by logging on to their
account and configuring their own preferences.

If staff need to be able to configure these preferences on behalf of
staged users, the "allow changing staged user tracking" site setting
can be enabled. Default is to not allow it.

### CODE REVIEW

File: app/serializers/web_hook_user_serializer.rb [hunk 1]
```diff
@@ -30,6 +30,7 @@ def staff_attributes(*attrs)
     can_change_bio
     can_change_location
     can_change_website
+    can_change_tracking_preferences
```
Comment (tgxworld) [inline app/serializers/web_hook_user_serializer.rb:33]: I don't think we should include this in the user web hook yet as this is doesn't really provide any benefit for the webhook.

File: lib/guardian/user_guardian.rb [hunk 2]
```diff
@@ -179,4 +179,9 @@ def can_upload_user_card_background?(user)
   def can_delete_sso_record?(user)
     SiteSetting.enable_discourse_connect && user && is_admin?
   end
+
+  def can_change_tracking_preferences?(user)
+    can_edit_user?(user) &&
+      (!user.staged || SiteSetting.allow_changing_staged_user_tracking)
```
Comment (tgxworld) [inline lib/guardian/user_guardian.rb:185]:
```suggestion
    (SiteSetting.allow_changing_staged_user_tracking || !user.staged) && can_edit_user?(user)
```

I flipped the conditions around based on which is cheaper to call.

File: spec/components/guardian/user_guardian_spec.rb [hunk 3]
```diff
@@ -455,4 +455,41 @@
       expect(guardian.can_upload_user_card_background?(trust_level_1)).to eq(false)
     end
   end
+
+  describe "can_change_tracking_preferences?" do
```
Comment (tgxworld) [inline spec/components/guardian/user_guardian_spec.rb:459]:
```suggestion
  describe "#can_change_tracking_preferences?" do
```

File: spec/components/guardian/user_guardian_spec.rb [hunk 4]
```diff
@@ -455,4 +455,41 @@
...
+    let(:staged_user) { Fabricate(:staged) }
+    let(:admin_user) { Fabricate(:admin) }
```
Comment (tgxworld) [inline spec/components/guardian/user_guardian_spec.rb:461]:
```suggestion
    fab!(:staged_user) { Fabricate(:staged) }
    fab!(:admin_user) { Fabricate(:admin) }
```

File: spec/services/user_updater_spec.rb [hunk 5]
```diff
@@ -72,7 +72,40 @@
...
       ).count).to eq(1)
+    end
+
+    context "staged user" do
+      let(:staged_user) { Fabricate(:staged) }
+
+      context "allow_changing_staged_user_tracking is false" do
+        before { SiteSetting.allow_changing_staged_user_tracking = false }
 
+        it "doesn't update categories and tags" do
```
Comment (tgxworld) [inline spec/services/user_updater_spec.rb:83]:
```suggestion
        it "doesn't update muted categories and watched tags" do
```

File: spec/services/user_updater_spec.rb [hunk 6]
```diff
@@ -72,7 +72,40 @@
...
+          updater = UserUpdater.new(Fabricate(:admin), staged_user)
+          updater.update(watched_tags: "#{tag.name}", muted_category_ids: [category.id])
+          expect(TagUser.where(user_id: staged_user.id).count).to eq(0)
+          expect(CategoryUser.where(user_id: staged_user.id).count).to eq(0)
```
Comment (tgxworld) [inline spec/services/user_updater_spec.rb:87]:
```suggestion
          expect(TagUser.exist?(user_id: staged_user.id)).to eq(false)
          expect(CategoryUser.exist?(user_id: staged_user.id)).to eq(false)
```

File: spec/services/user_updater_spec.rb [hunk 7]
```diff
@@ -72,7 +72,40 @@
...
+        end
+      end
+
+      context "allow_changing_staged_user_tracking is true" do
+        before { SiteSetting.allow_changing_staged_user_tracking = true }
+
+        it "updates categories and tags" do
```
Comment (tgxworld) [inline spec/services/user_updater_spec.rb:94]:
```suggestion
        it "updates muted categories and watched tags" do
```

File: spec/services/user_updater_spec.rb [hunk 8]
```diff
@@ -72,7 +72,40 @@
...
+            tag_id: tag.id,
+            notification_level: TagUser.notification_levels[:watching]
+          ).exists?).to eq(true)
+          expect(CategoryUser.where(
+            user_id: staged_user.id,
+            category_id: category.id,
+            notification_level: CategoryUser.notification_levels[:muted]
+          ).exists?).to eq(true)
+        end
+      end
```
Comment (tgxworld) [inline spec/services/user_updater_spec.rb:108]:
```suggestion
          expect(TagUser.exists?(
            user_id: staged_user.id,
            tag_id: tag.id,
            notification_level: TagUser.notification_levels[:watching]
          )).to eq(true)
          
          expect(CategoryUser.exists?(
            user_id: staged_user.id,
            category_id: category.id,
            notification_level: CategoryUser.notification_levels[:muted]
          )).to eq(true)
        end
      end
```

**Review by tgxworld — COMMENTED** [review]
I left some suggestions but in general the PR looks good to me.

File: spec/components/guardian/user_guardian_spec.rb (hunk 4 above)
Comment (nlalonde) [reply spec/components/guardian/user_guardian_spec.rb:461]: All tests don't need both users.

File: app/serializers/web_hook_user_serializer.rb (hunk 1 above)
Comment (nlalonde) [reply app/serializers/web_hook_user_serializer.rb]: Actually this code *was* removing it from the serializer. It is excluding attrs from the base serializer. I'll put it back.

File: app/assets/javascripts/discourse/app/controllers/preferences.js [hunk 9]
```diff
@@ -1,2 +1,12 @@
+import { and, readOnly } from "@ember/object/computed";
 import Controller from "@ember/controller";
-export default Controller.extend({});
+
+export default Controller.extend({
+  canChangeCategoryPreferences: readOnly(
+    "model.can_change_tracking_preferences"
+  ),
+  canChangeTagPreferences: and(
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/controllers/preferences.js:8]: Minor but if you are only using this in a template you can use the `(and` helper now which will eliminate some code.

**Review by eviltrout — APPROVED** [review]

File: app/assets/javascripts/discourse/app/controllers/preferences.js (hunk 9 above)
Comment (nlalonde) [reply app/assets/javascripts/discourse/app/controllers/preferences.js]: I was aware that was possible, but couldn't find an example of it. TIL it's `(and`.

**Pull request merged**